### ISSUE TITLE
Open questions recurrent policy

### ISSUE THREAD

**Issue by MarcoMeter** [MarcoMeter/neroRL#1]
## Assumptions

- Taining data has to be processed sequentially
- Gradients should not be backpropagated through data that is not part of the to be trained 
sequence

## Open Questions

1. Shall the training data be arranged in sequences of fixed length?

2. Is it sufficient to arrange the data into their respective episodes? (Well, that could cause inconsistent mini batch sizes or an episode might be incomplete)

3. What if a sequence ends up being shorter due to episode termination?

4. If those sequences are padded to maintain the fixed sequence length, doesn’t that increase the overall batch size (i.e. number of training samples)?

5. Upon optimizing the model, is it smarter to recompute the hidden states as they are modified by the optimization? (the originally hidden states from the sampling phase might be deprecated already after processing the first mini batch)

6. What if a sequence/episode was not completed during the data sampling phase?

7. How can you check in PyTorch how far the gradients are being backpropagated through the hidden states in the training data?

```
from torchviz import make_dot

graph = make_dot(final_tensor)
graph.view()
```

8. If sequences of fixed length are being trained, do you have to reset the hidden state for doing inference?

**Comment by MarcoMeter** [comment]
1. Yes
2. Depends on the environment
3. Padding or feed sequence by sequence (batch sequences of same length)
4. Padding causes higher memory usage
5. Not tested, but this has potential
6. Don't bother with that
7. Force torchviz to run on CPU to get the correct graph
8. No
